From d0401296addd55335ecdda2e40efc35253e7a8ce Mon Sep 17 00:00:00 2001
From: Abhay Krishna Arunachalam <arnchlm@amazon.com>
Date: Thu, 13 Mar 2025 12:11:42 -0700
Subject: [PATCH 13/16] Add non-EFI builds for RHEL 9 OVA

RHEL 9 OVA builds were defaulted to EFI upstream with no direct to override the firmware to BIOS. So we
are splitting the builds into BIOS and EFI builds so that customers can choose the firmware option based
on their requirements.

Signed-off-by: Abhay Krishna Arunachalam <arnchlm@amazon.com>
---
 images/capi/Makefile                          |  5 +-
 .../ova/linux/rhel/http/9/ks-efi.cfg.tmpl     | 82 +++++++++++++++++++
 .../packer/ova/linux/rhel/http/9/ks.cfg.tmpl  |  2 +-
 images/capi/packer/ova/rhel-9-efi.json        | 21 +++++
 images/capi/packer/ova/rhel-9.json            |  5 +-
 5 files changed, 110 insertions(+), 5 deletions(-)
 create mode 100644 images/capi/packer/ova/linux/rhel/http/9/ks-efi.cfg.tmpl
 create mode 100644 images/capi/packer/ova/rhel-9-efi.json

diff --git a/images/capi/Makefile b/images/capi/Makefile
index b2c354d75..362172a29 100644
--- a/images/capi/Makefile
+++ b/images/capi/Makefile
@@ -335,7 +335,7 @@ PACKER_POWERVS_NODE_FLAGS := $(foreach f,$(abspath $(COMMON_POWERVS_VAR_FILES)),
 CENTOS_VERSIONS			:=	centos-7
 FLATCAR_VERSIONS		:=	flatcar
 PHOTON_VERSIONS			:=	photon-3 photon-4 photon-5
-RHEL_VERSIONS			:=	rhel-7 rhel-8 rhel-9
+RHEL_VERSIONS			:=	rhel-7 rhel-8 rhel-9 rhel-9-efi
 ROCKYLINUX_VERSIONS		:=  rockylinux-8 rockylinux-9
 UBUNTU_VERSIONS			:=	ubuntu-2004 ubuntu-2004-efi ubuntu-2204 ubuntu-2204-efi ubuntu-2404 ubuntu-2404-efi
 WINDOWS_VERSIONS		:=	windows-2019 windows-2019-efi windows-2022 windows-2022-efi
@@ -757,6 +757,7 @@ build-node-ova-vsphere-photon-5: ## Builds Photon 5 Node OVA and template on vSp
 build-node-ova-vsphere-rhel-7: ## Builds RHEL 7 Node OVA and template on vSphere
 build-node-ova-vsphere-rhel-8: ## Builds RHEL 8 Node OVA and template on vSphere
 build-node-ova-vsphere-rhel-9: ## Builds RHEL 9 Node OVA and template on vSphere
+build-node-ova-vsphere-rhel-9-efi: ## Builds RHEL 9 Node OVA and template on vSphere that EFI boots
 build-node-ova-vsphere-rockylinux-8: ## Builds RockyLinux 8 Node OVA and template on vSphere
 build-node-ova-vsphere-rockylinux-9: ## Builds RockyLinux 9 Node OVA and template on vSphere
 build-node-ova-vsphere-ubuntu-2004: ## Builds Ubuntu 20.04 Node OVA and template on vSphere
@@ -778,6 +779,7 @@ build-node-ova-vsphere-clone-photon-5: ## Builds Photon 5 Node OVA and template
 build-node-ova-vsphere-clone-rhel-7: ## Builds RHEL 7 Node OVA and template on vSphere
 build-node-ova-vsphere-clone-rhel-8: ## Builds RHEL 8 Node OVA and template on vSphere
 build-node-ova-vsphere-clone-rhel-9: ## Builds RHEL 9 Node OVA and template on vSphere
+build-node-ova-vsphere-clone-rhel-9-efi: ## Builds RHEL 9 Node OVA and template on vSphere that EFI boots
 build-node-ova-vsphere-clone-rockylinux-8: ## Builds RockyLinux 8 Node OVA and template on vSphere
 build-node-ova-vsphere-clone-rockylinux-9: ## Builds RockyLinux 9 Node OVA and template on vSphere
 build-node-ova-vsphere-clone-ubuntu-2004: ## Builds Ubuntu 20.04 Node OVA and template on vSphere
@@ -794,6 +796,7 @@ build-node-ova-vsphere-base-photon-5: ## Builds base Photon 5 Node OVA and templ
 build-node-ova-vsphere-base-rhel-7: ## Builds base RHEL 7 Node OVA and template on vSphere
 build-node-ova-vsphere-base-rhel-8: ## Builds base RHEL 8 Node OVA and template on vSphere
 build-node-ova-vsphere-base-rhel-9: ## Builds base RHEL 9 Node OVA and template on vSphere
+build-node-ova-vsphere-base-rhel-9-efi: ## Builds base RHEL 9 Node OVA and template on vSphere that EFI boots
 build-node-ova-vsphere-base-rockylinux-8: ## Builds base RockyLinux 8 Node OVA and template on vSphere
 build-node-ova-vsphere-base-rockylinux-9: ## Builds base RockyLinux 9 Node OVA and template on vSphere
 build-node-ova-vsphere-base-ubuntu-2004: ## Builds base Ubuntu 20.04 Node OVA and template on vSphere
diff --git a/images/capi/packer/ova/linux/rhel/http/9/ks-efi.cfg.tmpl b/images/capi/packer/ova/linux/rhel/http/9/ks-efi.cfg.tmpl
new file mode 100644
index 000000000..cc13d548d
--- /dev/null
+++ b/images/capi/packer/ova/linux/rhel/http/9/ks-efi.cfg.tmpl
@@ -0,0 +1,82 @@
+# version=RHEL9
+# Use CDROM installation media
+cdrom
+
+# Use text install
+text
+
+# Don't run the Setup Agent on first boot
+firstboot --disabled
+eula --agreed
+
+# Keyboard layout
+keyboard --vckeymap=us --xlayouts='us'
+
+# System language
+lang en_US.UTF-8
+
+# Network information
+network --bootproto=dhcp --onboot=on --ipv6=auto --activate --hostname=rhel9
+
+# Lock Root account
+rootpw --lock
+
+# Create builder user
+user --name=builder --groups=wheel --password=$SSH_PASSWORD --plaintext --shell=/bin/bash
+
+# System services
+selinux --permissive
+firewall --disabled
+services --enabled=sshd
+
+# System timezone
+timezone UTC
+
+# System booloader configuration
+bootloader --location=mbr --boot-drive=sda
+zerombr
+clearpart --all --initlabel --drives=sda
+autopart --nohome --noswap --nolvm
+
+skipx
+
+%packages --ignoremissing --excludedocs
+# dnf group info minimal-environment
+@^minimal-environment
+@core
+openssh-server
+sed
+sudo
+python3
+open-vm-tools
+
+# Exclude unnecessary firmwares
+-iwl*firmware
+%end
+
+%addon com_redhat_kdump --disable
+%end
+
+reboot
+
+%post
+
+echo 'builder ALL=(ALL) NOPASSWD: ALL' >/etc/sudoers.d/builder
+chmod 440 /etc/sudoers.d/builder
+
+# Disable quiet boot and splash screen
+sed -i 's/ rhgb quiet//' /etc/default/grub
+sed -i 's/ rhgb quiet//' /boot/grub2/grubenv
+
+# Remove the package cache
+yum -y clean all
+
+# Disable swap
+swapoff -a
+rm -f /swapfile
+sed -ri '/\sswap\s/s/^#?/#/' /etc/fstab
+
+# Ensure on next boot that network devices get assigned unique IDs.
+sed -i '/^\(HWADDR\|UUID\)=/d' /etc/sysconfig/network-scripts/ifcfg-*
+
+%end
diff --git a/images/capi/packer/ova/linux/rhel/http/9/ks.cfg.tmpl b/images/capi/packer/ova/linux/rhel/http/9/ks.cfg.tmpl
index cc13d548d..6cc9a6ba1 100644
--- a/images/capi/packer/ova/linux/rhel/http/9/ks.cfg.tmpl
+++ b/images/capi/packer/ova/linux/rhel/http/9/ks.cfg.tmpl
@@ -36,7 +36,7 @@ timezone UTC
 bootloader --location=mbr --boot-drive=sda
 zerombr
 clearpart --all --initlabel --drives=sda
-autopart --nohome --noswap --nolvm
+part / --fstype="ext4" --grow --asprimary --label=slash --ondisk=sda
 
 skipx
 
diff --git a/images/capi/packer/ova/rhel-9-efi.json b/images/capi/packer/ova/rhel-9-efi.json
new file mode 100644
index 000000000..c5cbd9587
--- /dev/null
+++ b/images/capi/packer/ova/rhel-9-efi.json
@@ -0,0 +1,21 @@
+{
+  "boot_command_prefix": "<up>e <down><down><end><wait> text inst.ks=",
+  "boot_command_suffix": "/9/ks-efi.cfg<enter><wait><leftCtrlOn>x<leftCtrlOff>",
+  "boot_media_path": "http://{{ .HTTPIP }}:{{ .HTTPPort }}",
+  "build_name": "rhel-9",
+  "distro_arch": "amd64",
+  "distro_name": "rhel",
+  "distro_version": "9",
+  "epel_rpm_gpg_key": "https://dl.fedoraproject.org/pub/epel/RPM-GPG-KEY-EPEL-9",
+  "firmware": "efi",
+  "guest_os_type": "rhel9-64",
+  "http_directory": "./packer/ova/linux/{{user `distro_name`}}/http/",
+  "iso_checksum": "398561d7b66f1a4bf23664f4aa8f2cfbb3641aa2f01a320068e86bd1fc0e9076",
+  "iso_checksum_type": "sha256",
+  "iso_url": "file:///rhel-9.4-x86_64-dvd.iso",
+  "os_display_name": "RHEL 9",
+  "redhat_epel_rpm": "https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm",
+  "shutdown_command": "shutdown -P now",
+  "vmx_version": "18",
+  "vsphere_guest_os_type": "rhel9_64Guest"
+}
diff --git a/images/capi/packer/ova/rhel-9.json b/images/capi/packer/ova/rhel-9.json
index 0b806d1a8..91e351f1a 100644
--- a/images/capi/packer/ova/rhel-9.json
+++ b/images/capi/packer/ova/rhel-9.json
@@ -1,13 +1,12 @@
 {
-  "boot_command_prefix": "<up>e <down><down><end><wait> text inst.ks=",
-  "boot_command_suffix": "/9/ks.cfg<enter><wait><leftCtrlOn>x<leftCtrlOff>",
+  "boot_command_prefix": "<up><tab> text inst.ks=",
+  "boot_command_suffix": "/9/ks.cfg<enter><wait>",
   "boot_media_path": "http://{{ .HTTPIP }}:{{ .HTTPPort }}",
   "build_name": "rhel-9",
   "distro_arch": "amd64",
   "distro_name": "rhel",
   "distro_version": "9",
   "epel_rpm_gpg_key": "https://dl.fedoraproject.org/pub/epel/RPM-GPG-KEY-EPEL-9",
-  "firmware": "efi",
   "guest_os_type": "rhel9-64",
   "http_directory": "./packer/ova/linux/{{user `distro_name`}}/http/",
   "iso_checksum": "398561d7b66f1a4bf23664f4aa8f2cfbb3641aa2f01a320068e86bd1fc0e9076",
-- 
2.49.0

